---
title: "Genomics Functions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{genomics-functions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(levinmisc)
```

<!-- WARNING - This vignette is generated by {fusen} from dev/flat_genomics_functions.Rmd: do not edit by hand -->

# Plotting

These convenience functions are useful for creating plots from GWAS summary statistics.


## Manhattan plots

The `gg_manhattan_df` function is used to create Manhattan plots. This function can optionally annotated the plot with additional information (eg. gene name, rsid) by using the `annotation_df` amd `label_col` arguments.
    

  

```{r example-gg_manhattan_df}
#' \dontrun{
#' gg_manhattan_df(sumstats_df)
#' }
```

  

  

## QQ plots

The `gg_qq_df` function can be used to generate a qq plot to visually evaluate for test statistic inflation, and calculate lambda GC to quantify any inflation.
    

  

```{r example-gg_qq_df}
#' \dontrun{
#' gg_qq_df(sumstats_df)
#' }
```

  

  
---


# Annotation

These convenience functions are useful for adding additional information to GWAS summary stastics.


## Add rsids

This function can be used rapidly add rsids to GWAS summary statistics by chromosome:position.
    

  

```{r example-annotate_rsids}
#' \dontrun{
#' annotate_rsids(sumstats_df)
#' }
```

  

  
---


# GWAS Meta-analysis

GWAS summary statistics from multiple studies can be combined using meta-analysis. The following functions are useful for orchestrating GWAS meta-analysis using R.


## METAL

METAL (<https://github.com/statgen/METAL>) is a common command-line tool for performing GWAS meta-analysis. The following functions can be used to configure and run METAL directly from R.


## Configure METAL

    

  

```{r example-metal_config}
#' \dontrun{
#' metal_config(config_name = "name-of-analysis", output_dir = "/path/to/output/", study_files = c("/path/to/sumstats_1.txt", "/path/to/sumstats_2.txt))
#' }
```

  

## Run METAL

    

  

```{r example-metal_run}
#' \dontrun{
#' metal_run(config_file = "config.txt", metal_path = "/path/to/metal_binary")
#' }
```

  

## Run MR-MEGA

    

  

```{r example-mr_mega}
#' \dontrun{
#' mr_mega(sumstats_files = list("/path/to/sumstats_1.txt.gz", "/path/to/sumstats_2.txt.gz), mr_mega_bin = "/path/to/MR-MEGA")
#' }
```

  

  

---


# Linkage Disequilibirium

## plink

The `plink_extract_ld` function is a wrapper around the `plink` v1.9 binary, which takes a list of variants and uses a reference panel to calculate the linkage disequilibrium among variants.

    

  

```{r example-plink_extract_ld}
#' \dontrun{
#' plink_extract_ld(df, snp_col, effect_allele_col, metric = "r", bfile = "/path/to/reference/panel/", plink_bin = "/path/to/plink1.9")
#' }
```

  

  

<!-- ```{r function-plink_extract_ld_possibly} -->
<!-- #' plink_extract_ld_possibly -->
<!-- #'  -->
<!-- #' An internal wrapper around plink_extract_ld_possibly to silently return `NULL` on error -->
<!-- #'  -->
<!-- #' @return -->
<!-- #'  -->
<!-- #' @noRd -->
<!-- plink_extract_ld_possibly <- purrr::possibly(levinmisc::plink_extract_ld) -->
<!-- ``` -->

<!-- ```{r example-plink_extract_ld_possibly} -->
<!-- #' \dontrun{ -->
<!-- #' plink_extract_ld_possibly() -->
<!-- #' } -->
<!-- ``` -->

<!-- ```{r tests-plink_extract_ld_possibly} -->
<!-- test_that("plink_extract_ld_possibly works", { -->
<!--   expect_true(inherits(plink_extract_ld_possibly, "function"))  -->
<!-- }) -->
<!-- ``` -->
  
---  
  

# Colocalization

Colocalization is a technique for evaluating the evidence supporting the presence of shared causal variant(s) at a given locus across two or more traits. Several methods of colocalization have been described, which generally leverage GWAS summary statistics across multiple traits. The methods use either proportional or enumeration approaches, which make different assumptions and test different hypotheses.


## HyPrColoc

This function is a convenience wrapper around [hyprcoloc::hyprcoloc()] that takes a dataframe as input, and performs mutli-trait colocalization using HyPrColoc, a Bayesian enumeration colocalization method which makes the assumption of a single causal variant at the locus. Details of the HyPrColoc method are described in Foley et al. (Nature Communications 2021; <https://doi.org/10.1038/s41467-020-20885-8>).


  

```{r example-hyprcoloc_df}
#' \dontrun{
#' hyprcoloc_df(gwas_results_df)
#' }
```

  

  


## Coloc

    

  

```{r example-coloc_run}
#' \dontrun{
#' coloc_run(locus_df)
#' }
```

  

  
  
  
---
  

# Finemapping

Finemapping is an approach for identifying the putative causal variant(s) at a locus identified in a GWAS. Like colocalization, finemapping methods make different assumptions about the configuration of causal variant(s) at the locus.


## Bayesian finemapping

This function implements the simple Bayesian finemapping approach described in Mallard et al. (Nature Genetics 2012; <https://doi.org/10.1038/ng.2435>), adapted in Graham et al. (Nature 2021; <https://doi.org/10.1038/s41586-021-04064-3>). This method assumes a single causal variant configuration at the locus.
    

  

```{r example-calc_credset}
#' \dontrun{
#' calc_credset(gwas_df)
#' }
```

  

## CARMA

This function implements the CARMA finemapping approach described in Yang et al. (Nature Genetics 2023; <https://doi.org/10.1038/s41588-023-01392-0>). This method relaxes the single causal variant assumption, identifying independent credible sets at each fine-mapped locus.
    

  

```{r example-run_carma}
#' \dontrun{
#' run_carma(locus_df, snp_col = SNP, z_col = z, effect_allele_col = allele1)
#' }
```

  

  

---    
  

# Gene-based Testing

GWAS summary statistics can be used to perform gene-based testing for associations with the trait/phenotype of interest, rather than SNP/variant-level associations found in GWAS.


## MAGMA

This function provides a wrapper to MAGMA, which performs gene-based testing from GWAS summary statistics. Details of MAGMA described in de Leeuw et al. (PLoS Cumputational Biology 2015; <https://doi.org/10.1371/journal.pcbi.1004219>). The MAGMA binary and reference files are available from the Complex Trait Genetics lab (<https://ctg.cncr.nl/software/magma>).
    

  

```{r example-magmar}
#' \dontrun{
#' magmar(sumstats_df, magma_bin = "/path/to/magma", bfile = "/path/to/bfiles", gene_file = "/path/to/genes", out_file = "magma_output")
#' }
```

  

    


## S-PrediXcan

    

  

```{r example-s_predixcan}
#' \dontrun{
#' s_predixcan(df, data = "MetaXcan/data", metaxcan = "MetaXcan/software", output = "/path/to/output", model_db_path = "MetaXcan/data/models/eqtl/mashr/mashr_Liver.db", model_covariance_path = "MetaXcan/data/models/eqtl/mashr/mashr_Liver.txt.gz", trait_name = "GWAS_trait")
#' }
```

  

      


## S-MultiXcan

    

  

```{r example-s_multixcan}
#' \dontrun{
#' s_multixcan(df, 
#'  models_folder = "MetaXcan/data/models/eqtl/mashr/",
#'  models_name_filter = ".*\\.db",
#'  models_name_pattern = "mashr_(.*)\\.db",
#'  metaxcan_folder = "/path/to/metaxcan_files/,
#'  metaxcan_filter = "GWAS-trait_mashr_(.*)_S-prediXcan.csv",
#'  metaxcan_file_name_parse_pattern = "(.*)_mashr_(.*)_S-prediXcan.csv",
#'  data = "MetaXcan/data",
#'  metaxcan = "MetaXcan/software",
#'  snp_covariance = "MetaXcan/data/models/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz")
#' }
```

  

  

---

# Heritability

## LDAK

    

  

```{r example-ldak_h2}
#' \dontrun{
#' ldak_h2(sumstats_file = "/path/to/munged_sumstats", ldak_bin = "/path/to/ldak", ldak_tagfile = "/path/to/tagfile",  hm3_file = "/path/to/hm3")
#' }
```

  

  


