---
title: "Genomics Functions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{genomics-functions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(levinmisc)
```

<!-- WARNING - This vignette is generated by {fusen} from /dev/flat_genomics_functions.Rmd: do not edit by hand -->

# Plotting

## Manhattan plots

The `gg_manhattan_df` function is used to create Manhattan plots. This function can optionally annotated the plot with additional information (eg. gene name, rsid) by using the `annotation_df` amd `label_col` arguments.
    

  

```{r example-gg_manhattan_df}
#' \dontrun{
#' gg_manhattan_df()
#' }
```

  

  

## QQ plots

The `gg_qq_df` function can be used to generate a qq plot to visually evaluate for test statistic inflation, and calculate lambda GC to quantify any inflation.
    

  

```{r example-gg_qq_df}
#' \dontrun{
#' gg_qq_df()
#' }
```

  

  

# Annotation

## Add rsids

This function can be used rapidly add rsids to GWAS summary statistics by chromosome:position.
    

  

```{r example-annotate_rsids}
#' \dontrun{
#' annotate_rsids()
#' }
```

  

  

# METAL

## Configure METAL

    

  

```{r example-metal_config}
#' \dontrun{
#' metal_config()
#' }
```

  

## Run METAL

    

  

```{r example-metal_run}
#' \dontrun{
#' metal_run(config_file = "config.txt", metal_path = "/path/to/metal_binary")
#' }
```

  

  
  

# Colocalization

## HyPrColoc

    

  

```{r example-hyprcoloc_df}
#' \dontrun{
#' hyprcoloc_df(gwas_results_df)
#' }
```

  

  


