<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ levinmisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="targets">Targets<a class="anchor" aria-label="anchor" href="#targets"></a></h2>
      
      <p class="section-desc"></p><p>Build a simple {targets} project</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="populate_targets_proj.html">populate_targets_proj()</a></code> 
        </dt>
        <dd>Create a minimal targets template in the current project</dd>
      </dl><dl><dt>
          
          <code><a href="use_crew_lsf.html">use_crew_lsf()</a></code> 
        </dt>
        <dd>Use Crew LSF</dd>
      </dl></div><div class="section level2">
      <h2 id="genomics">Genomics<a class="anchor" aria-label="anchor" href="#genomics"></a></h2>
      
      <p class="section-desc"></p><p>Genomics utility functions and wrappers around other tools/programs/packages to perform GWAS meta-analysis, fine-mapping, colocalization, and more.</p>

      
    </div><div class="section level2">
      
      <h3 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for annotating GWAS summary statistics</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annotate_rsids.html">annotate_rsids()</a></code> 
        </dt>
        <dd>Annotate a dataframe containing genomic coordinates with rsids</dd>
      </dl></div><div class="section level2">
      
      <h3 id="gwas">GWAS<a class="anchor" aria-label="anchor" href="#gwas"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for performing GWAS/GWAS meta-analysis</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="metal_config.html">metal_config()</a></code> 
        </dt>
        <dd>Create a configuration file for METAL</dd>
      </dl><dl><dt>
          
          <code><a href="metal_run.html">metal_run()</a></code> 
        </dt>
        <dd>Use METAL to run a GWAS meta-analysis</dd>
      </dl><dl><dt>
          
          <code><a href="mr_mega.html">mr_mega()</a></code> 
        </dt>
        <dd>Perform multi-ancestry GWAS meta-analysis using MR-MEGA</dd>
      </dl></div><div class="section level2">
      
      <h3 id="colocalization">Colocalization<a class="anchor" aria-label="anchor" href="#colocalization"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for performing colocalization. Colocalization is a technique for evaluating the evidence supporting the presence of shared causal variant(s) at a given locus across two or more traits. Several methods of colocalization have been described, which generally leverage GWAS summary statistics across multiple traits. The methods use either proportional or enumeration approaches, which make different assumptions and test different hypotheses.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="coloc_run.html">coloc_run()</a></code> 
        </dt>
        <dd>Run Bayesian enumeration colocalization using Coloc</dd>
      </dl><dl><dt>
          
          <code><a href="hyprcoloc_df.html">hyprcoloc_df()</a></code> 
        </dt>
        <dd>Run multi-trait colocalization using HyPrColoc</dd>
      </dl></div><div class="section level2">
      
      <h3 id="finemapping">Finemapping<a class="anchor" aria-label="anchor" href="#finemapping"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for finemapping. Finemapping is an approach for identifying the putative causal variant(s) at a locus identified in a GWAS. Like colocalization, finemapping methods make different assumptions about the configuration of causal variant(s) at the locus.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="calc_credset.html">calc_credset()</a></code> 
        </dt>
        <dd>Perform Bayesian finemapping</dd>
      </dl></div><div class="section level2">
      
      <h3 id="heritability">Heritability<a class="anchor" aria-label="anchor" href="#heritability"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for estimating heritability. Several tools for performing heritability estimation using GWAS summary statistics have been developed, including LDSC and LDAK. The <a href="https://mglev1n.github.io/ldscr/" class="external-link">ldscr</a> package provides a native R implemtation of LDSC.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="ldak_h2.html">ldak_h2()</a></code> 
        </dt>
        <dd>Calculate heritability using LDAK</dd>
      </dl></div><div class="section level2">
      
      <h3 id="twas-gene-based-testing">TWAS/Gene-based testing<a class="anchor" aria-label="anchor" href="#twas-gene-based-testing"></a></h3>
      <p class="section-desc"></p><p>Functions that are useful for identifying trait-associated genes from GWAS.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="s_multixcan.html">s_multixcan()</a></code> 
        </dt>
        <dd>Integrate PrediXcan data across tissues</dd>
      </dl><dl><dt>
          
          <code><a href="s_predixcan.html">s_predixcan()</a></code> 
        </dt>
        <dd>Run a TWAS using S-PrediXcan</dd>
      </dl><dl><dt>
          
          <code><a href="magmar.html">magmar()</a></code> 
        </dt>
        <dd>Run MAGMA gene-based analysis</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a></h2>
      
      <p class="section-desc"></p><p>Miscellaneous functions useful for manipulating/presenting data.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="render_datatable.html">render_datatable()</a></code> 
        </dt>
        <dd>Render DataTable to HTML</dd>
      </dl><dl><dt>
          
          <code><a href="gg_manhattan_df.html">gg_manhattan_df()</a></code> 
        </dt>
        <dd>Create a Manhattan Plot</dd>
      </dl><dl><dt>
          
          <code><a href="gg_qq_df.html">gg_qq_df()</a></code> 
        </dt>
        <dd>Create a QQ plot</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

