<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a configuration file for METAL — metal_config • levinmisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a configuration file for METAL — metal_config"><meta name="description" content="This function can be used to generate a configuration for METAL, a tool for performing meta-analysis of GWAS summary statistics https://github.com/statgen/METAL. The file created by this function can be used to run a meta-analysis. Details of the arguments to METAL are described in the METAL documentation: https://genome.sph.umich.edu/wiki/METAL_Documentation."><meta property="og:description" content="This function can be used to generate a configuration for METAL, a tool for performing meta-analysis of GWAS summary statistics https://github.com/statgen/METAL. The file created by this function can be used to run a meta-analysis. Details of the arguments to METAL are described in the METAL documentation: https://genome.sph.umich.edu/wiki/METAL_Documentation."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a configuration file for METAL</h1>
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/levinmisc/blob/main/R/metal_config.R" class="external-link"><code>R/metal_config.R</code></a></small>
      <div class="d-none name"><code>metal_config.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function can be used to generate a configuration for METAL, a tool for performing meta-analysis of GWAS summary statistics <a href="https://github.com/statgen/METAL" class="external-link">https://github.com/statgen/METAL</a>. The file created by this function can be used to run a meta-analysis. Details of the arguments to METAL are described in the METAL documentation: <a href="https://genome.sph.umich.edu/wiki/METAL_Documentation" class="external-link">https://genome.sph.umich.edu/wiki/METAL_Documentation</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">metal_config</span><span class="op">(</span></span>
<span>  <span class="va">config_name</span>,</span>
<span>  <span class="va">output_dir</span>,</span>
<span>  <span class="va">study_files</span>,</span>
<span>  SCHEME <span class="op">=</span> <span class="st">"STDERR"</span>,</span>
<span>  AVERAGEFREQ <span class="op">=</span> <span class="st">"ON"</span>,</span>
<span>  MINMAXFREQ <span class="op">=</span> <span class="st">"OFF"</span>,</span>
<span>  TRACKPOSITIONS <span class="op">=</span> <span class="st">"ON"</span>,</span>
<span>  MARKERLABEL <span class="op">=</span> <span class="st">"MARKER"</span>,</span>
<span>  CHROMOSOMELABEL <span class="op">=</span> <span class="st">"CHROM"</span>,</span>
<span>  POSITIONLABEL <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>  EFFECT_ALLELE <span class="op">=</span> <span class="st">"EFFECT_ALLELE"</span>,</span>
<span>  OTHER_ALLELE <span class="op">=</span> <span class="st">"OTHER_ALLELE"</span>,</span>
<span>  EFFECTLABEL <span class="op">=</span> <span class="st">"BETA"</span>,</span>
<span>  STDERR <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>  FREQLABEL <span class="op">=</span> <span class="st">"EAF"</span>,</span>
<span>  NCASE <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  NCONTROL <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  SAMPLESIZE <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-config-name">config_name<a class="anchor" aria-label="anchor" href="#arg-config-name"></a></dt>
<dd><p>(string) Name of the configuration</p></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>(path) Path to output directory where GWAS meta-analysis data should be stored</p></dd>


<dt id="arg-study-files">study_files<a class="anchor" aria-label="anchor" href="#arg-study-files"></a></dt>
<dd><p>(vector) Character vector of paths to summary statistics that should be included in the GWAS</p></dd>


<dt id="arg-scheme">SCHEME<a class="anchor" aria-label="anchor" href="#arg-scheme"></a></dt>
<dd><p>(string) Either "SAMPLESIZE" or "STDERR" (default), corresponding to the METAL analysis scheme</p></dd>


<dt id="arg-averagefreq">AVERAGEFREQ<a class="anchor" aria-label="anchor" href="#arg-averagefreq"></a></dt>
<dd><p>(string) Either "ON" (default) or "OFF", allowing METAL to report the mean effect allele frequency across files</p></dd>


<dt id="arg-minmaxfreq">MINMAXFREQ<a class="anchor" aria-label="anchor" href="#arg-minmaxfreq"></a></dt>
<dd><p>(string) Either "ON" (default) or "OFF", allowing METAL to report the min/max effect allele frequency across files</p></dd>


<dt id="arg-trackpositions">TRACKPOSITIONS<a class="anchor" aria-label="anchor" href="#arg-trackpositions"></a></dt>
<dd><p>(string) Either "ON" (default) or "OFF", allowing METAL to report chromosome/position in the output</p></dd>


<dt id="arg-markerlabel">MARKERLABEL<a class="anchor" aria-label="anchor" href="#arg-markerlabel"></a></dt>
<dd><p>(string) Column containing unique markers to analyze (this column must be named the same across all input files)</p></dd>


<dt id="arg-chromosomelabel">CHROMOSOMELABEL<a class="anchor" aria-label="anchor" href="#arg-chromosomelabel"></a></dt>
<dd><p>(string) Column containing chromosomes (this column must be named the same across all input files)</p></dd>


<dt id="arg-positionlabel">POSITIONLABEL<a class="anchor" aria-label="anchor" href="#arg-positionlabel"></a></dt>
<dd><p>(string) Column containing genomic positions (this column must be named the same across all input files)</p></dd>


<dt id="arg-effect-allele">EFFECT_ALLELE<a class="anchor" aria-label="anchor" href="#arg-effect-allele"></a></dt>
<dd><p>(string) Column containing effect alleles (this column must be named the same across all input files)</p></dd>


<dt id="arg-other-allele">OTHER_ALLELE<a class="anchor" aria-label="anchor" href="#arg-other-allele"></a></dt>
<dd><p>(string) Column containing non-effect alleles (this column must be named the same across all input files)</p></dd>


<dt id="arg-effectlabel">EFFECTLABEL<a class="anchor" aria-label="anchor" href="#arg-effectlabel"></a></dt>
<dd><p>(string) Column containing effect sizes corresponding to the effect allele (this column must be named the same across all input files)</p></dd>


<dt id="arg-stderr">STDERR<a class="anchor" aria-label="anchor" href="#arg-stderr"></a></dt>
<dd><p>(string) Column containing standard errors fo the effect estimate (this column must be named the same across all input files)</p></dd>


<dt id="arg-freqlabel">FREQLABEL<a class="anchor" aria-label="anchor" href="#arg-freqlabel"></a></dt>
<dd><p>(string) Column containing effect allele frequencies (this column must be named the same across all input files)</p></dd>


<dt id="arg-ncase">NCASE<a class="anchor" aria-label="anchor" href="#arg-ncase"></a></dt>
<dd><p>(string) Column containing number of cases (this column must be named the same across all input files)</p></dd>


<dt id="arg-ncontrol">NCONTROL<a class="anchor" aria-label="anchor" href="#arg-ncontrol"></a></dt>
<dd><p>(string) Column containing number of controls (this column must be named the same across all input files)</p></dd>


<dt id="arg-samplesize">SAMPLESIZE<a class="anchor" aria-label="anchor" href="#arg-samplesize"></a></dt>
<dd><p>(string) Column containing total samplesize (this column must be named the same across all input files)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Path to METAL configuration file</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Run GWAS meta-analysis using METAL: <code><a href="metal_run.html">metal_run()</a></code></p>
<p>Other GWAS meta-analysis:
<code><a href="metal_run.html">metal_run</a>()</code>,
<code><a href="mr_mega.html">mr_mega</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">metal_config</span><span class="op">(</span>config_name <span class="op">=</span> <span class="st">"name-of-analysis"</span>, output_dir <span class="op">=</span> <span class="st">"/path/to/output/"</span>, study_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/path/to/sumstats_1.txt"</span>, <span class="st">"/path/to/sumstats_2.txt"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

