<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform Bayesian finemapping using CARMA — run_carma • levinmisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Bayesian finemapping using CARMA — run_carma"><meta name="description" content="This function is a wrapper around CARMA::CARMA() that takes a dataframe containing variants at a locus and performs Bayesian finemapping based on their signed z-scores and an LD reference panel. The function requires a plink-formatted LD reference panel (bfile), which will be used to generate a signed LD matrix at the locus. CARMA was described in Yang et al. (Nature Genetics 2023; https://doi.org/10.1038/s41588-023-01392-0)"><meta property="og:description" content="This function is a wrapper around CARMA::CARMA() that takes a dataframe containing variants at a locus and performs Bayesian finemapping based on their signed z-scores and an LD reference panel. The function requires a plink-formatted LD reference panel (bfile), which will be used to generate a signed LD matrix at the locus. CARMA was described in Yang et al. (Nature Genetics 2023; https://doi.org/10.1038/s41588-023-01392-0)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform Bayesian finemapping using CARMA</h1>
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/levinmisc/blob/main/R/run_carma.R" class="external-link"><code>R/run_carma.R</code></a></small>
      <div class="d-none name"><code>run_carma.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper around <code><a href="https://rdrr.io/pkg/CARMA/man/CARMA.html" class="external-link">CARMA::CARMA()</a></code> that takes a dataframe containing variants at a locus and performs Bayesian finemapping based on their signed z-scores and an LD reference panel. The function requires a plink-formatted LD reference panel (<code>bfile</code>), which will be used to generate a signed LD matrix at the locus. CARMA was described in Yang et al. (Nature Genetics 2023; <a href="https://doi.org/10.1038/s41588-023-01392-0" class="external-link">https://doi.org/10.1038/s41588-023-01392-0</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_carma</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">snp_col</span>,</span>
<span>  <span class="va">z_col</span>,</span>
<span>  <span class="va">effect_allele_col</span>,</span>
<span>  outlier_switch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">bfile</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  memory <span class="op">=</span> <span class="fl">16000</span>,</span>
<span>  <span class="va">plink_bin</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>Dataframe containing variants at a locus for finemapping</p></dd>


<dt id="arg-snp-col">snp_col<a class="anchor" aria-label="anchor" href="#arg-snp-col"></a></dt>
<dd><p>Name of column containing SNP identifiers</p></dd>


<dt id="arg-z-col">z_col<a class="anchor" aria-label="anchor" href="#arg-z-col"></a></dt>
<dd><p>Name of column containing signed Z-scores (relative to effect allele)</p></dd>


<dt id="arg-effect-allele-col">effect_allele_col<a class="anchor" aria-label="anchor" href="#arg-effect-allele-col"></a></dt>
<dd><p>Name of column containing effect alleles</p></dd>


<dt id="arg-outlier-switch">outlier_switch<a class="anchor" aria-label="anchor" href="#arg-outlier-switch"></a></dt>
<dd><p>Whether to remove outliers when performing finemapping. Argument is passed to the <code>outlier.switch</code> argument of <code><a href="https://rdrr.io/pkg/CARMA/man/CARMA.html" class="external-link">CARMA::CARMA()</a></code></p></dd>


<dt id="arg-bfile">bfile<a class="anchor" aria-label="anchor" href="#arg-bfile"></a></dt>
<dd><p>Path to plink <code>bfile</code> of reference panel that will be used to extract LD</p></dd>


<dt id="arg-threads">threads<a class="anchor" aria-label="anchor" href="#arg-threads"></a></dt>
<dd><p>Number of threads (default = 1)</p></dd>


<dt id="arg-memory">memory<a class="anchor" aria-label="anchor" href="#arg-memory"></a></dt>
<dd><p>Memory limit (default = 16000 MB)</p></dd>


<dt id="arg-plink-bin">plink_bin<a class="anchor" aria-label="anchor" href="#arg-plink-bin"></a></dt>
<dd><p>Path to plink executable</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional arguments passsed to <code><a href="https://rdrr.io/pkg/CARMA/man/CARMA.html" class="external-link">CARMA::CARMA()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe containing the input dataframe, and additional columns denoting which credible set (<code>CS</code>) each variant belongs to, as well as the posterior inclusion probability (<code>PIP</code>), and an <code>ld_error</code> column noting whether there were problems generating the LD matrix that limited fine-mapping.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other finemapping:
<code><a href="calc_credset.html">calc_credset</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">run_carma</span><span class="op">(</span><span class="va">locus_df</span>, snp_col <span class="op">=</span> <span class="va">SNP</span>, z_col <span class="op">=</span> <span class="va">z</span>, effect_allele_col <span class="op">=</span> <span class="va">allele1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

