<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by s_predixcan(). The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; https://doi.org/10.1371/journal.pgen.1007889). The MetaXcan tools can be found on Github (https://github.com/hakyimlab/MetaXcan) and PredictDB (https://predictdb.org/)."><title>Integrate PrediXcan data across tissues — s_multixcan • levinmisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integrate PrediXcan data across tissues — s_multixcan"><meta property="og:description" content="This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by s_predixcan(). The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; https://doi.org/10.1371/journal.pgen.1007889). The MetaXcan tools can be found on Github (https://github.com/hakyimlab/MetaXcan) and PredictDB (https://predictdb.org/)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a>
    <a class="dropdown-item" href="../articles/genomics-functions.html">Genomics Functions</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Integrate PrediXcan data across tissues</h1>
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/levinmisc/blob/HEAD/R/s_multixcan.R" class="external-link"><code>R/s_multixcan.R</code></a></small>
      <div class="d-none name"><code>s_multixcan.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by <code><a href="s_predixcan.html">s_predixcan()</a></code>. The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; <a href="https://doi.org/10.1371/journal.pgen.1007889" class="external-link">https://doi.org/10.1371/journal.pgen.1007889</a>). The MetaXcan tools can be found on Github (<a href="https://github.com/hakyimlab/MetaXcan" class="external-link">https://github.com/hakyimlab/MetaXcan</a>) and PredictDB (<a href="https://predictdb.org/" class="external-link">https://predictdb.org/</a>).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">s_multixcan</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">models_folder</span>,</span>
<span>  models_name_filter <span class="op">=</span> <span class="st">".*db"</span>,</span>
<span>  <span class="va">models_name_pattern</span>,</span>
<span>  snp <span class="op">=</span> <span class="va">SNP</span>,</span>
<span>  effect_allele <span class="op">=</span> <span class="va">effect_allele</span>,</span>
<span>  other_allele <span class="op">=</span> <span class="va">other_allele</span>,</span>
<span>  beta <span class="op">=</span> <span class="va">beta</span>,</span>
<span>  eaf <span class="op">=</span> <span class="va">eaf</span>,</span>
<span>  chr <span class="op">=</span> <span class="va">chr</span>,</span>
<span>  pos <span class="op">=</span> <span class="va">pos</span>,</span>
<span>  se <span class="op">=</span> <span class="va">se</span>,</span>
<span>  pval <span class="op">=</span> <span class="va">pval</span>,</span>
<span>  samplesize <span class="op">=</span> <span class="va">samplesize</span>,</span>
<span>  regularization <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  cutoff_condition_number <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  cutoff_eigen_ratio <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  cutoff_threshold <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  cutoff_trace_ratio <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  <span class="va">metaxcan_folder</span>,</span>
<span>  <span class="va">metaxcan_filter</span>,</span>
<span>  <span class="va">metaxcan_file_name_parse_pattern</span>,</span>
<span>  <span class="va">snp_covariance</span>,</span>
<span>  <span class="va">metaxcan</span>,</span>
<span>  <span class="va">data</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>df</dt>
<dd><p>Dataframe containing GWAS summary statistics</p></dd>


<dt>models_folder</dt>
<dd><p>Path to folder containing MetaXcan models (eg. <code>"MetaXcan/data/models/eqtl/mashr/"</code>)</p></dd>


<dt>models_name_filter</dt>
<dd><p>Filter used to identify model files (default = <code>".*\\.db"</code>)</p></dd>


<dt>models_name_pattern</dt>
<dd><p>Filter used to extract trait name from the model (default = <code>"mashr_(.*)\\.db"</code>)</p></dd>


<dt>snp</dt>
<dd><p>Column containing rsids</p></dd>


<dt>effect_allele</dt>
<dd><p>Column containing effect alleles</p></dd>


<dt>other_allele</dt>
<dd><p>Column containing non-effect alleles</p></dd>


<dt>beta</dt>
<dd><p>Column containing effect estimates</p></dd>


<dt>eaf</dt>
<dd><p>Column containing effect allele frequencies</p></dd>


<dt>chr</dt>
<dd><p>Column containing chromosomes</p></dd>


<dt>pos</dt>
<dd><p>Column containing positions</p></dd>


<dt>se</dt>
<dd><p>Column containing standard errors</p></dd>


<dt>pval</dt>
<dd><p>Column containing p-values</p></dd>


<dt>samplesize</dt>
<dd><p>Column containing samplesize</p></dd>


<dt>regularization</dt>
<dd><p>MetaXcan regularization (default = 0.01)</p></dd>


<dt>cutoff_condition_number</dt>
<dd><p>Numer of eigen values to use when truncating SVD components (default = 30)</p></dd>


<dt>cutoff_eigen_ratio</dt>
<dd><p>Ratio of eigenvalues to the max eigenvalue, as threshold to use when truncating SVD components. (default = 0.001)</p></dd>


<dt>cutoff_threshold</dt>
<dd><p>Threshold of variance eigenvalues when truncating SVD (default = 0.4)</p></dd>


<dt>cutoff_trace_ratio</dt>
<dd><p>Ratio of eigenvalues to trace, to use when truncating SVD (default = 0.01)</p></dd>


<dt>metaxcan_folder</dt>
<dd><p>Path to folder containing S-PrediXcan result files</p></dd>


<dt>metaxcan_filter</dt>
<dd><p>Regular expression to filter results files</p></dd>


<dt>metaxcan_file_name_parse_pattern</dt>
<dd><p>Regular expression to get phenotype name and model name from MetaXcan result files. Assumes that a first group will be matched to phenotype name, and the second to model name.</p></dd>


<dt>snp_covariance</dt>
<dd><p>Path to file containing S-MultiXcan covariance (eg. <code>"MetaXcan/data/models/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz"</code>)</p></dd>


<dt>metaxcan</dt>
<dd><p>Path to MetaXcan (eg. <code>"MetaXcan/software"</code>)</p></dd>


<dt>data</dt>
<dd><p>Path to MetaXcan data (eg. <code>"MetaXcan/data"</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A dataframe containing S-MultiXcan results</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other TWAS: 
<code><a href="s_predixcan.html">s_predixcan</a>()</code></p>
<p>Other Gene-based testing: 
<code><a href="magmar.html">magmar</a>()</code>,
<code><a href="s_predixcan.html">s_predixcan</a>()</code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

