<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Integrate PrediXcan data across tissues — s_multixcan • levinmisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integrate PrediXcan data across tissues — s_multixcan"><meta name="description" content="This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by s_predixcan(). The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; https://doi.org/10.1371/journal.pgen.1007889). The MetaXcan tools can be found on Github (https://github.com/hakyimlab/MetaXcan) and PredictDB (https://predictdb.org/)."><meta property="og:description" content="This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by s_predixcan(). The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; https://doi.org/10.1371/journal.pgen.1007889). The MetaXcan tools can be found on Github (https://github.com/hakyimlab/MetaXcan) and PredictDB (https://predictdb.org/)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Integrate PrediXcan data across tissues</h1>
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/levinmisc/blob/main/R/s_multixcan.R" class="external-link"><code>R/s_multixcan.R</code></a></small>
      <div class="d-none name"><code>s_multixcan.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper around S-MultiXcan, a tool for identifying genes associate with a trait using GWAS summary statistics. This approach leverages the sharing of eQTLs acorss tissues to improve upon tissue-specific TWAS-methods like S-PrediXcan. This method uses data generated by <code><a href="s_predixcan.html">s_predixcan()</a></code>. The S-MultiXcan method was described in Barbeira et al. (PLOS Genetics 2019; <a href="https://doi.org/10.1371/journal.pgen.1007889" class="external-link">https://doi.org/10.1371/journal.pgen.1007889</a>). The MetaXcan tools can be found on Github (<a href="https://github.com/hakyimlab/MetaXcan" class="external-link">https://github.com/hakyimlab/MetaXcan</a>) and PredictDB (<a href="https://predictdb.org/" class="external-link">https://predictdb.org/</a>).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">s_multixcan</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">models_folder</span>,</span>
<span>  models_name_filter <span class="op">=</span> <span class="st">".*db"</span>,</span>
<span>  <span class="va">models_name_pattern</span>,</span>
<span>  snp <span class="op">=</span> <span class="va">SNP</span>,</span>
<span>  effect_allele <span class="op">=</span> <span class="va">effect_allele</span>,</span>
<span>  other_allele <span class="op">=</span> <span class="va">other_allele</span>,</span>
<span>  beta <span class="op">=</span> <span class="va">beta</span>,</span>
<span>  eaf <span class="op">=</span> <span class="va">eaf</span>,</span>
<span>  chr <span class="op">=</span> <span class="va">chr</span>,</span>
<span>  pos <span class="op">=</span> <span class="va">pos</span>,</span>
<span>  se <span class="op">=</span> <span class="va">se</span>,</span>
<span>  pval <span class="op">=</span> <span class="va">pval</span>,</span>
<span>  samplesize <span class="op">=</span> <span class="va">samplesize</span>,</span>
<span>  regularization <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  cutoff_condition_number <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  cutoff_eigen_ratio <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  cutoff_threshold <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  cutoff_trace_ratio <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  <span class="va">metaxcan_folder</span>,</span>
<span>  <span class="va">metaxcan_filter</span>,</span>
<span>  <span class="va">metaxcan_file_name_parse_pattern</span>,</span>
<span>  <span class="va">snp_covariance</span>,</span>
<span>  <span class="va">metaxcan</span>,</span>
<span>  <span class="va">data</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>Dataframe containing GWAS summary statistics</p></dd>


<dt id="arg-models-folder">models_folder<a class="anchor" aria-label="anchor" href="#arg-models-folder"></a></dt>
<dd><p>Path to folder containing MetaXcan models (eg. <code>"MetaXcan/data/models/eqtl/mashr/"</code>)</p></dd>


<dt id="arg-models-name-filter">models_name_filter<a class="anchor" aria-label="anchor" href="#arg-models-name-filter"></a></dt>
<dd><p>Filter used to identify model files (default = <code>".*\\.db"</code>)</p></dd>


<dt id="arg-models-name-pattern">models_name_pattern<a class="anchor" aria-label="anchor" href="#arg-models-name-pattern"></a></dt>
<dd><p>Filter used to extract trait name from the model (default = <code>"mashr_(.*)\\.db"</code>)</p></dd>


<dt id="arg-snp">snp<a class="anchor" aria-label="anchor" href="#arg-snp"></a></dt>
<dd><p>Column containing rsids</p></dd>


<dt id="arg-effect-allele">effect_allele<a class="anchor" aria-label="anchor" href="#arg-effect-allele"></a></dt>
<dd><p>Column containing effect alleles</p></dd>


<dt id="arg-other-allele">other_allele<a class="anchor" aria-label="anchor" href="#arg-other-allele"></a></dt>
<dd><p>Column containing non-effect alleles</p></dd>


<dt id="arg-beta">beta<a class="anchor" aria-label="anchor" href="#arg-beta"></a></dt>
<dd><p>Column containing effect estimates</p></dd>


<dt id="arg-eaf">eaf<a class="anchor" aria-label="anchor" href="#arg-eaf"></a></dt>
<dd><p>Column containing effect allele frequencies</p></dd>


<dt id="arg-chr">chr<a class="anchor" aria-label="anchor" href="#arg-chr"></a></dt>
<dd><p>Column containing chromosomes</p></dd>


<dt id="arg-pos">pos<a class="anchor" aria-label="anchor" href="#arg-pos"></a></dt>
<dd><p>Column containing positions</p></dd>


<dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Column containing standard errors</p></dd>


<dt id="arg-pval">pval<a class="anchor" aria-label="anchor" href="#arg-pval"></a></dt>
<dd><p>Column containing p-values</p></dd>


<dt id="arg-samplesize">samplesize<a class="anchor" aria-label="anchor" href="#arg-samplesize"></a></dt>
<dd><p>Column containing samplesize</p></dd>


<dt id="arg-regularization">regularization<a class="anchor" aria-label="anchor" href="#arg-regularization"></a></dt>
<dd><p>MetaXcan regularization (default = 0.01)</p></dd>


<dt id="arg-cutoff-condition-number">cutoff_condition_number<a class="anchor" aria-label="anchor" href="#arg-cutoff-condition-number"></a></dt>
<dd><p>Numer of eigen values to use when truncating SVD components (default = 30)</p></dd>


<dt id="arg-cutoff-eigen-ratio">cutoff_eigen_ratio<a class="anchor" aria-label="anchor" href="#arg-cutoff-eigen-ratio"></a></dt>
<dd><p>Ratio of eigenvalues to the max eigenvalue, as threshold to use when truncating SVD components. (default = 0.001)</p></dd>


<dt id="arg-cutoff-threshold">cutoff_threshold<a class="anchor" aria-label="anchor" href="#arg-cutoff-threshold"></a></dt>
<dd><p>Threshold of variance eigenvalues when truncating SVD (default = 0.4)</p></dd>


<dt id="arg-cutoff-trace-ratio">cutoff_trace_ratio<a class="anchor" aria-label="anchor" href="#arg-cutoff-trace-ratio"></a></dt>
<dd><p>Ratio of eigenvalues to trace, to use when truncating SVD (default = 0.01)</p></dd>


<dt id="arg-metaxcan-folder">metaxcan_folder<a class="anchor" aria-label="anchor" href="#arg-metaxcan-folder"></a></dt>
<dd><p>Path to folder containing S-PrediXcan result files</p></dd>


<dt id="arg-metaxcan-filter">metaxcan_filter<a class="anchor" aria-label="anchor" href="#arg-metaxcan-filter"></a></dt>
<dd><p>Regular expression to filter results files</p></dd>


<dt id="arg-metaxcan-file-name-parse-pattern">metaxcan_file_name_parse_pattern<a class="anchor" aria-label="anchor" href="#arg-metaxcan-file-name-parse-pattern"></a></dt>
<dd><p>Regular expression to get phenotype name and model name from MetaXcan result files. Assumes that a first group will be matched to phenotype name, and the second to model name.</p></dd>


<dt id="arg-snp-covariance">snp_covariance<a class="anchor" aria-label="anchor" href="#arg-snp-covariance"></a></dt>
<dd><p>Path to file containing S-MultiXcan covariance (eg. <code>"MetaXcan/data/models/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz"</code>)</p></dd>


<dt id="arg-metaxcan">metaxcan<a class="anchor" aria-label="anchor" href="#arg-metaxcan"></a></dt>
<dd><p>Path to MetaXcan (eg. <code>"MetaXcan/software"</code>)</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>Path to MetaXcan data (eg. <code>"MetaXcan/data"</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe containing S-MultiXcan results</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other TWAS:
<code><a href="s_predixcan.html">s_predixcan</a>()</code></p>
<p>Other Gene-based testing:
<code><a href="magmar.html">magmar</a>()</code>,
<code><a href="s_predixcan.html">s_predixcan</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">s_multixcan</span><span class="op">(</span><span class="va">df</span>,</span></span>
<span class="r-in"><span>  models_folder <span class="op">=</span> <span class="st">"MetaXcan/data/models/eqtl/mashr/"</span>,</span></span>
<span class="r-in"><span>  models_name_filter <span class="op">=</span> <span class="st">".*\\.db"</span>,</span></span>
<span class="r-in"><span>  models_name_pattern <span class="op">=</span> <span class="st">"mashr_(.*)\\.db"</span>,</span></span>
<span class="r-in"><span>  metaxcan_folder <span class="op">=</span> <span class="st">"/path/to/metaxcan_files/"</span>,</span></span>
<span class="r-in"><span>  metaxcan_filter <span class="op">=</span> <span class="st">"GWAS-trait_mashr_(.*)_S-prediXcan.csv"</span>,</span></span>
<span class="r-in"><span>  metaxcan_file_name_parse_pattern <span class="op">=</span> <span class="st">"(.*)_mashr_(.*)_S-prediXcan.csv"</span>,</span></span>
<span class="r-in"><span>  data <span class="op">=</span> <span class="st">"MetaXcan/data"</span>,</span></span>
<span class="r-in"><span>  metaxcan <span class="op">=</span> <span class="st">"MetaXcan/software"</span>,</span></span>
<span class="r-in"><span>  snp_covariance <span class="op">=</span> <span class="st">"MetaXcan/data/models/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

