<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="levinmisc">
<title>Genomics Functions • levinmisc</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genomics Functions">
<meta property="og:description" content="levinmisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">levinmisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create-a-new-targets-pipeline-for-lpc.html">Create a New Targets Pipeline for LPC</a>
    <a class="dropdown-item" href="../articles/genomics-functions.html">Genomics Functions</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mglev1n/levinmisc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Genomics Functions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/levinmisc/blob/HEAD/vignettes/genomics-functions.Rmd" class="external-link"><code>vignettes/genomics-functions.Rmd</code></a></small>
      <div class="d-none name"><code>genomics-functions.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mglev1n/levinmisc" class="external-link">levinmisc</a></span><span class="op">)</span></span></code></pre></div>
<!-- WARNING - This vignette is generated by {fusen} from dev/flat_genomics_functions.Rmd: do not edit by hand -->
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<p>These convenience functions are useful for creating plots from GWAS
summary statistics.</p>
<div class="section level3">
<h3 id="manhattan-plots">Manhattan plots<a class="anchor" aria-label="anchor" href="#manhattan-plots"></a>
</h3>
<p>The <code>gg_manhattan_df</code> function is used to create Manhattan
plots. This function can optionally annotated the plot with additional
information (eg. gene name, rsid) by using the
<code>annotation_df</code> amd <code>label_col</code> arguments.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' gg_manhattan_df(sumstats_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="qq-plots">QQ plots<a class="anchor" aria-label="anchor" href="#qq-plots"></a>
</h3>
<p>The <code>gg_qq_df</code> function can be used to generate a qq plot
to visually evaluate for test statistic inflation, and calculate lambda
GC to quantify any inflation.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' gg_qq_df(sumstats_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h2>
<p>These convenience functions are useful for adding additional
information to GWAS summary stastics.</p>
<div class="section level3">
<h3 id="add-rsids">Add rsids<a class="anchor" aria-label="anchor" href="#add-rsids"></a>
</h3>
<p>This function can be used rapidly add rsids to GWAS summary
statistics by chromosome:position.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' annotate_rsids(sumstats_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="gwas-meta-analysis">GWAS Meta-analysis<a class="anchor" aria-label="anchor" href="#gwas-meta-analysis"></a>
</h2>
<p>GWAS summary statistics from multiple studies can be combined using
meta-analysis. The following functions are useful for orchestrating GWAS
meta-analysis using R.</p>
<div class="section level3">
<h3 id="metal">METAL<a class="anchor" aria-label="anchor" href="#metal"></a>
</h3>
<p>METAL (<a href="https://github.com/statgen/METAL" class="external-link uri">https://github.com/statgen/METAL</a>) is a common
command-line tool for performing GWAS meta-analysis. The following
functions can be used to configure and run METAL directly from R.</p>
</div>
<div class="section level3">
<h3 id="configure-metal">Configure METAL<a class="anchor" aria-label="anchor" href="#configure-metal"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' metal_config(config_name = "name-of-analysis", output_dir = "/path/to/output/", study_files = c("/path/to/sumstats_1.txt", "/path/to/sumstats_2.txt))</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-metal">Run METAL<a class="anchor" aria-label="anchor" href="#run-metal"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' metal_run(config_file = "config.txt", metal_path = "/path/to/metal_binary")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-mr-mega">Run MR-MEGA<a class="anchor" aria-label="anchor" href="#run-mr-mega"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' mr_mega(sumstats_files = list("/path/to/sumstats_1.txt.gz", "/path/to/sumstats_2.txt.gz), mr_mega_bin = "/path/to/MR-MEGA")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="linkage-disequilibirium">Linkage Disequilibirium<a class="anchor" aria-label="anchor" href="#linkage-disequilibirium"></a>
</h2>
<div class="section level3">
<h3 id="plink">plink<a class="anchor" aria-label="anchor" href="#plink"></a>
</h3>
<p>The <code>plink_extract_ld</code> function is a wrapper around the
<code>plink</code> v1.9 binary, which takes a list of variants and uses
a reference panel to calculate the linkage disequilibrium among
variants.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' plink_extract_ld(df, snp_col, effect_allele_col, metric = "r", bfile = "/path/to/reference/panel/", plink_bin = "/path/to/plink1.9")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<!-- ```{r function-plink_extract_ld_possibly} -->
<!-- #' plink_extract_ld_possibly -->
<!-- #'  -->
<!-- #' An internal wrapper around plink_extract_ld_possibly to silently return `NULL` on error -->
<!-- #'  -->
<!-- #' @return -->
<!-- #'  -->
<!-- #' @noRd -->
<!-- plink_extract_ld_possibly <- purrr::possibly(levinmisc::plink_extract_ld) -->
<!-- ``` -->
<!-- ```{r example-plink_extract_ld_possibly} -->
<!-- #' \dontrun{ -->
<!-- #' plink_extract_ld_possibly() -->
<!-- #' } -->
<!-- ``` -->
<!-- ```{r tests-plink_extract_ld_possibly} -->
<!-- test_that("plink_extract_ld_possibly works", { -->
<!--   expect_true(inherits(plink_extract_ld_possibly, "function"))  -->
<!-- }) -->
<!-- ``` -->
<hr>
</div>
</div>
<div class="section level2">
<h2 id="colocalization">Colocalization<a class="anchor" aria-label="anchor" href="#colocalization"></a>
</h2>
<p>Colocalization is a technique for evaluating the evidence supporting
the presence of shared causal variant(s) at a given locus across two or
more traits. Several methods of colocalization have been described,
which generally leverage GWAS summary statistics across multiple traits.
The methods use either proportional or enumeration approaches, which
make different assumptions and test different hypotheses.</p>
<div class="section level3">
<h3 id="hyprcoloc">HyPrColoc<a class="anchor" aria-label="anchor" href="#hyprcoloc"></a>
</h3>
<p>This function is a convenience wrapper around
[hyprcoloc::hyprcoloc()] that takes a dataframe as input, and performs
mutli-trait colocalization using HyPrColoc, a Bayesian enumeration
colocalization method which makes the assumption of a single causal
variant at the locus. Details of the HyPrColoc method are described in
Foley et al. (Nature Communications 2021; <a href="https://doi.org/10.1038/s41467-020-20885-8" class="external-link uri">https://doi.org/10.1038/s41467-020-20885-8</a>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' hyprcoloc_df(gwas_results_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="coloc">Coloc<a class="anchor" aria-label="anchor" href="#coloc"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' coloc_run(locus_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="finemapping">Finemapping<a class="anchor" aria-label="anchor" href="#finemapping"></a>
</h2>
<p>Finemapping is an approach for identifying the putative causal
variant(s) at a locus identified in a GWAS. Like colocalization,
finemapping methods make different assumptions about the configuration
of causal variant(s) at the locus.</p>
<div class="section level3">
<h3 id="bayesian-finemapping">Bayesian finemapping<a class="anchor" aria-label="anchor" href="#bayesian-finemapping"></a>
</h3>
<p>This function implements the simple Bayesian finemapping approach
described in Mallard et al. (Nature Genetics 2012; <a href="https://doi.org/10.1038/ng.2435" class="external-link uri">https://doi.org/10.1038/ng.2435</a>), adapted in Graham et
al. (Nature 2021; <a href="https://doi.org/10.1038/s41586-021-04064-3" class="external-link uri">https://doi.org/10.1038/s41586-021-04064-3</a>). This method
assumes a single causal variant configuration at the locus.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' calc_credset(gwas_df)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="carma">CARMA<a class="anchor" aria-label="anchor" href="#carma"></a>
</h3>
<p>This function implements the CARMA finemapping approach described in
Yang et al. (Nature Genetics 2023; <a href="https://doi.org/10.1038/s41588-023-01392-0" class="external-link uri">https://doi.org/10.1038/s41588-023-01392-0</a>). This method
relaxes the single causal variant assumption, identifying independent
credible sets at each fine-mapped locus.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' run_carma(locus_df, snp_col = SNP, z_col = z, effect_allele_col = allele1)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="gene-based-testing">Gene-based Testing<a class="anchor" aria-label="anchor" href="#gene-based-testing"></a>
</h2>
<p>GWAS summary statistics can be used to perform gene-based testing for
associations with the trait/phenotype of interest, rather than
SNP/variant-level associations found in GWAS.</p>
<div class="section level3">
<h3 id="magma">MAGMA<a class="anchor" aria-label="anchor" href="#magma"></a>
</h3>
<p>This function provides a wrapper to MAGMA, which performs gene-based
testing from GWAS summary statistics. Details of MAGMA described in de
Leeuw et al. (PLoS Cumputational Biology 2015; <a href="https://doi.org/10.1371/journal.pcbi.1004219" class="external-link uri">https://doi.org/10.1371/journal.pcbi.1004219</a>). The MAGMA
binary and reference files are available from the Complex Trait Genetics
lab (<a href="https://ctg.cncr.nl/software/magma" class="external-link uri">https://ctg.cncr.nl/software/magma</a>).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' magmar(sumstats_df, magma_bin = "/path/to/magma", bfile = "/path/to/bfiles", gene_file = "/path/to/genes", out_file = "magma_output")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="s-predixcan">S-PrediXcan<a class="anchor" aria-label="anchor" href="#s-predixcan"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' s_predixcan(df, data = "MetaXcan/data", metaxcan = "MetaXcan/software", output = "/path/to/output", model_db_path = "MetaXcan/data/models/eqtl/mashr/mashr_Liver.db", model_covariance_path = "MetaXcan/data/models/eqtl/mashr/mashr_Liver.txt.gz", trait_name = "GWAS_trait")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="s-multixcan">S-MultiXcan<a class="anchor" aria-label="anchor" href="#s-multixcan"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' s_multixcan(df, </span></span>
<span><span class="co">#'  models_folder = "MetaXcan/data/models/eqtl/mashr/",</span></span>
<span><span class="co">#'  models_name_filter = ".*\\.db",</span></span>
<span><span class="co">#'  models_name_pattern = "mashr_(.*)\\.db",</span></span>
<span><span class="co">#'  metaxcan_folder = "/path/to/metaxcan_files/,</span></span>
<span><span class="co">#'  metaxcan_filter = "GWAS-trait_mashr_(.*)_S-prediXcan.csv",</span></span>
<span><span class="co">#'  metaxcan_file_name_parse_pattern = "(.*)_mashr_(.*)_S-prediXcan.csv",</span></span>
<span><span class="co">#'  data = "MetaXcan/data",</span></span>
<span><span class="co">#'  metaxcan = "MetaXcan/software",</span></span>
<span><span class="co">#'  snp_covariance = "MetaXcan/data/models/gtex_v8_expression_mashr_snp_smultixcan_covariance.txt.gz")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="heritability">Heritability<a class="anchor" aria-label="anchor" href="#heritability"></a>
</h2>
<div class="section level3">
<h3 id="ldak">LDAK<a class="anchor" aria-label="anchor" href="#ldak"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co">#' ldak_h2(sumstats_file = "/path/to/munged_sumstats", ldak_bin = "/path/to/ldak", ldak_tagfile = "/path/to/tagfile",  hm3_file = "/path/to/hm3")</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
