# WARNING - Generated by {fusen} from dev/flat_genomics_functions.Rmd: do not edit by hand # nolint: line_length_linter.

#' Perform Bayesian finemapping using the Approximate Bayes Factor approach
#'
#' Description
#'
#' @param df Dataframe containing GWAS summary statistics
#' @param locus_marker_col Column containing a locus-level identifier
#' @param effect_col Column containing effect estimates
#' @param se_col Column containing standard errors fo the effect estimates
#' @param samplesize_col Column containing sample sizes
#' @param cred_interval Credible interval for the fine-mapped credible sets (default = 0.99; 0.95 is another common but artbitrarily determined interval)
#'
#' @return A data.frame containing credible sets at each locus. For each variant within the credible set, the prior probability of being the casual variant is provided.
#'
#' @export
#' @family {finemapping}
#' @concept genomics
#' @examples
#' \dontrun{
#' calc_credset(gwas_df)
#' }

calc_credset <- function(df, locus_marker_col = locus_marker, effect_col = effect, se_col = std_err, samplesize_col = samplesize, cred_interval = 0.99) {
  df %>%
    dplyr::group_by({{ locus_marker_col }}) %>%
    dplyr::mutate(bf = exp(0.5 * ({{ effect_col }}^2 / {{ se_col }}^2 - log({{ samplesize_col }})))) %>%
    dplyr::mutate(posterior_prob = bf / sum(bf)) %>%
    dplyr::arrange(dplyr::desc(posterior_prob)) %>%
    dplyr::mutate(cum_sum = cumsum(posterior_prob)) %>%
    dplyr::group_by({{ locus_marker_col }}) %>%
    dplyr::filter(cum_sum <= cred_interval | posterior_prob > cred_interval) %>%
    dplyr::select(-cum_sum) %>%
    dplyr::ungroup()
}
