# WARNING - Generated by {fusen} from dev/flat_genomics_functions.Rmd: do not edit by hand

#' Run MAGMA gene-based analysis
#' 
#' This function provides a wrapper to MAGMA, which performs gene-based testing from GWAS summary statistics. Details of MAGMA described in de Leeuw et al. (PLoS Cumputational Biology 2015; <https://doi.org/10.1371/journal.pcbi.1004219>). The MAGMA binary and reference files are available from the Complex Trait Genetics lab (<https://ctg.cncr.nl/software/magma>).
#' 
#' @param sumstats_df Dataframe containing GWAS summary statistics
#' @param pval_col Column containing p-values
#' @param snp_col Column containing rsids
#' @param samplesize_col Column containing sample size
#' @param magma_bin Path to MAGMA binary
#' @param bfile Path to reference data in plink `bfile` format (path should not include the extensions)
#' @param gene_file Path to file containing gene locations
#' @param out_file Output directory + file prefix
#'
#' @return Path to MAGMA results file
#' @concept genomics
#' @family Gene-based testing
#' @export
#' 
#' @examples
#' \dontrun{
#' magmar(sumstats_df, magma_bin = "/path/to/magma", bfile = "/path/to/bfiles", gene_file = "/path/to/genes", out_file = "magma_output")
#' }
magmar <- function(sumstats_df, pval_col = p_value, snp_col = rsid, samplesize_col = samplesize, magma_bin, bfile, gene_file, out_file) {
  cli::cli_progress_step("Writing summary statistics to temporary file")
  sumstats_file <- fs::file_temp()
  sumstats_df %>%
    select(SNP = {{ snp_col }}, P = {{ pval_col }}, N = {{ samplesize_col }}) %>%
    vroom::vroom_write(sumstats_file)


  cli::cli_progress_step("Running MAGMA")

  # create output path if it doesn't exist
  fs::dir_create(fs::path_dir(out_file))

  processx::run(magma_bin,
    args = c(
      "--bfile", bfile,
      "--gene-annot", gene_file,
      "--pval", sumstats_file, "ncol=N",
      "--gene-model", "snp-wise=mean",
      "--out", out_file
    ),
    error_on_status = FALSE,
    echo_cmd = TRUE,
    echo = TRUE
  )

  genes_out <- paste0(out_file, ".genes.out")
  genes_raw <- paste0(out_file, ".genes.raw")

  return(
    c("genes_out" = genes_out)
  )
}
