# WARNING - Generated by {fusen} from dev/flat_genomics_functions.Rmd: do not edit by hand

#' Extract an LD matrix from a reference panel using plink 1.9
#'
#' @param df Dataframe containing variants to be included in the LD matrix; must contain SNP identifier and effect allele
#' @param snp_col Name of column containing SNP identifiers
#' @param effect_allele_col Name of column containing effect alleles (used when considering signed measures of LD like `r`)
#' @param metric LD metric - either `r` (which will return a signed correlation matrix) or `r2` (which will return squared correlations)
#' @param bfile Path to plink `bfile` that will be used to extract LD
#' @param threads Number of threads (default = 1)
#' @param memory Memory limit (default = 16000 MB)
#' @param plink_bin Path to plink executable
#'
#' @return A square matrix containing the pairwise correlations between genetic variants
#' @export
#' @concept genomics
#' @family {linkage disequilibrium}
#'
#' @examples
#' \dontrun{
#' plink_extract_ld(df, snp_col, effect_allele_col, metric = "r", bfile = "/path/to/reference/panel/", plink_bin = "/path/to/plink1.9")
#' }
plink_extract_ld <- function(df, snp_col, effect_allele_col, metric = "r", bfile, threads = 1, memory = 16000,
                             plink_bin) {
  snp_id <- df %>%
    select(SNP = {{ snp_col }}) %>%
    head(1) %>%
    pull(SNP)

  snp_file <- fs::file_temp(pattern = snp_id)

  ld_dir <- fs::path_temp(snp_id)

  fs::dir_create(ld_dir, recurse = TRUE)

  df %>%
    select(SNP = {{ snp_col }}, snp_ideffect_allele = {{ effect_allele_col }}) %>%
    write_tsv(snp_file, col_names = FALSE)

  cli::cli_alert_info("Extracting LD")

  processx::run(plink_bin,
    args = c(
      "--threads", threads,
      "--memory", 16000,
      "--bfile", bfile,
      "--extract", snp_file,
      "--a1-allele", snp_file,
      "2 1 '#'",
      paste0("--", metric), "square", "gz",
      "--out", fs::path(ld_dir, "LD")
    ),
    error_on_status = FALSE,
    echo_cmd = TRUE,
    echo = TRUE
  )

  ld_matrix <- data.table::fread(fs::path(ld_dir, "LD.ld.gz"), col.names = df %>% pull({{ snp_col }})) %>%
    as.matrix()

  rownames(ld_matrix) <- df %>% pull({{ snp_col }})

  return(ld_matrix)
}
